### ISSUE TITLE
OVM static endpoint

### ISSUE THREAD

**Issue by davidvossel** [kubevirt/kubevirt#861]
OVMs are meant to be immortal, but the underlying tech that supports OVMs (pods) are mortal. Between each OVM start/stop/start sequence, a new pod is created to host the VM.  This means that the same OVM will have multiple pod IPs throughout it's lifecycle.

Users who want access a OVM using an address that is static and consistent between restarts will have to couple OVMs with a k8s service using a ClusterIP. This requires manually maintaining both the individual OVM object and the Service object.

If this relationship between OVM and k8s service is desirable enough, we should consider providing a shortcut that allows users to specific ports that they want to be statically accessible throughout the OVMs lifecycle. The OVM controller would then automatically generate the service endpoint for the OVM.

Example: Create an OVM and specify that you want static ssh and rdp access. Behind the scenes the OVM controller creates a ClusterIP that can be used to access those ports.  The ClusterIP will be accessible in the OVMs status field after creation.
```
apiVersion: kubevirt.io/v1alpha1
kind: OfflineVirtualMachine
metadata:
  name: testvm
spec:
  running: true
  ports:
    - port: 22
      targetPort: 22
      protocol: TCP
    - port: 3389
      targetPort: 3389
      protocol: TCP
  template:
    spec:
      domain:
        resources:
          requests:
            memory: 64M
        devices:
          disks:
            - name: registrydisk
              volumeName: registryvolume
              disk:
                bus: virtio
      volumes:
        - name: registryvolume
          registryDisk:
            image: kubevirt/cirros-registry-disk-demo:devel
```

**Comment by itamarh** [comment]
ClusterIP is a virtual allocation, right?
Its an interesting approach, but I don't know it will solve the way VMs communicate to other services/VMs, live migration (not supposed to break on-going communication, etc. - I think with Cluster IP a live migration would break on going ssh sessions?).
(I wonder if we wouldn't need to find a way to get the IP allocation logic to allow re-using IP's at cluster level for pod's, a-la dhcp, etc. - I know it sounds strange in the context of kubernetes)

**Comment by rmohr** [comment]
> OVMs are meant to be immortal, but the underlying tech that supports OVMs (pods) are mortal. Between each OVM start/stop/start sequence, a new pod is created to host the VM. This means that the same OVM will have multiple pod IPs throughout it's lifecycle.

My take would be to use kube-dns and not try to provide something static regarding to IPs. It should imo be completely fine if one decides to recreate a service which sits in front of a vm with and which gets a new cluster IP.

> Users who want access a OVM using an address that is static and consistent between restarts will have to couple OVMs with a k8s service using a ClusterIP. This requires manually maintaining both the individual OVM object and the Service object.

I don't think that we need a static ClusterIP for that. I think that inside the cluster kube-dns should take care of that. I guess the static IP only makes sense from an ingress point of view.

> If this relationship between OVM and k8s service is desirable enough, we should consider providing a shortcut that allows users to specific ports that they want to be statically accessible throughout the OVMs lifecycle. The OVM controller would then automatically generate the service endpoint for the OVM.

I think it highly depends on how you manage your VMs and what they are supposed to do. Crating a service will always be possible, no matter if the vm exists or not (and we know the unique labels the vm pod would get), so I think we are fine with plane services here in general. However I would agree that we might want some more comfort for the user who wants to expose special "admin" endpoints like ssh or rdp (real vm ports) or vnc and spice.

Depending on the use-case we have different possibilities to make life for people easier.

* In-Cluster examples:
 - virtctl expose vm/ovm myvm --port=22 --name=myvmfrontend
 - headless service with subdomains and name (see also #784) which provide direct dns entries per vm
 - even having a dedicated VirtualMachineService may be an option (especially when thinking about migrations).
* External examples:
 - running e.g. ansible on our vms (ingress with external ip/port per vm, host ports?)
 - virtctl port-forward myvm 5000 6000
 - virtctl ssh myvm (could look for a specific port name by default. E.g. have a port named SSH on your vm/pod)

**Comment by rmohr** [comment]
> ClusterIP is a virtual allocation, right?
> Its an interesting approach, but I don't know it will solve the way VMs communicate to other services/VMs, live migration (not supposed to break on-going communication, etc. - I think with Cluster IP a live migration would break on going ssh sessions?).

Indeed.

 (I wonder if we wouldn't need to find a way to get the IP allocation logic to allow re-using IP's at cluster level for pod's, a-la dhcp, etc. - I know it sounds strange in the context of kubernetes)

Hm, that part is left to the implementation of each cni plugin. E.g. flannel assigns IP ranges per host. With ovn it might even be possible to somehow reuse them, however I am not sure if we would want to go that route. Another option is to have something like a VirtualMachineService, which works pretty much like a Service, except that it would give us the abstraction to integrate the VM interfaces and ips into kube-dns our own way (k8s has already everything in place to integrate services with their endpoints framework). We could use CNI then, but we don't have to, and could do our own ip logic. However, I am not sure, how much routing/iptables magic is still happening in the kubelet, so there might be hidden pitfalls too.

**Comment by davidvossel** [comment]
> My take would be to use kube-dns and not try to provide something static regarding to IPs.

DNS caching and TTLs... In a situation where people are running their own VMs, we don't have control over how either of those things are performed by the client. This leads to situations where clients are using stale DNS entries and are unable to connect to the OVM after a restart. 

I've seen several cloud providers that provide a concept of a Cluster IP that can be assigned to a VM as a static endpoint. I've also seen that some people even choose to create a load balancer with only a single endpoint behind it to solve this problem of having a static endpoint to reference as well. 

This is a good conversation :) I think the time is right for us to begin coming up with our plan for this

**Comment by rmohr** [comment]
> DNS caching and TTLs... In a situation where people are running their own VMs, we don't have control over how either of those things are performed by the client. This leads to situations where clients are using stale DNS entries and are unable to connect to the OVM after a restart.

That sounds indeed tricky to solve. We can keep the Service in place when the vm gets shut down. That would keep the clusterIP then (but not sure if this kind of ClusterIP would really meet all our requirements). For instance, as you mentioned offline that would not allow non-TCP/UDP communication between vms.

As I see it right now, if we have the same ttl inside the vm like we have right now for pods, our vms will have the same chance to reach other services like pods have right now, so from an application point of view that should be at least sufficient for a lot of use-cases. Also pods will have the same guarantee to reach services within the vms like they have right now.

> This is a good conversation :) I think the time is right for us to begin coming up with our plan for this

Yes, a lot of cases to think through ...

**Comment by fabiand** [comment]
Service provide us with a static endpoint mechanism which works for VMs and OVMs - although there might be a small gap around labels.
Live migration is ignored in this context - but might be addressed with the multinetwork/l2 work we do on a different side.

tl;dr

As we do not have it - and because the current Kubernetes network model does not support it - I am ignoring requirements for live migration right now. but it's not forgotten.

To me the guiding principle in this discussion is how we achieve stable addressing of workloads in kubernetes. IOW - For the user it should be pretty similar - workflow and technical wise - how a VM and how a pod is getting exposed.

In Kube you can speak directly to pods - which is unreliable as the pods themselfes - or more reliable (with a static endpoint) if the user is creating a service.
Thus for KubeVirt we should adopt the same:
- VMs are exposed by using Services to expose certain ports.
- OVMs are exposed by using Services to expose certain ports of the corresponding VMs.

Here we just need to make sure that a user can choose between exposing the port of a VM _instance_ (VM) or a _logical_ VM (OVM). This should work depending on the selector the user uses for the service:

- OVM: Use the domain in the selector
- VM: UUID of the VM? There might be a gap here.

> If this relationship between OVM and k8s service is desirable enough, we should consider providing a shortcut that allows users to specific ports that they want to be statically accessible throughout the OVMs lifecycle. The OVM controller would then automatically generate the service endpoint for the OVM.

In Kubernetes this could have been done for i.e. Deployments as well, but so far this did not happen, thus I'd also follow this approach and not do to much magic.

**Comment by fabiand** [comment]
Just leaving a note - The important bit is that labels provide sufficient information to identify _all_ instances of an OVm over time.
If the labels are stable, then Sevrces can easily be written to address all VM instances (over time) even if the instances change - which is logcially equivalent to letting a Service point to an OVM.

**Comment by rmohr** [comment]
Picking up the discussion. To summarize my thoughts, we have:

 * W have the same reachability guarantee through kube-dns backed services for pods like we have for VMs
 * We have `virtctl expose <myvm>` which works on VMs, to make life for people easy
 * Labels on the VMI in the VM template work like they work on pods regarding to services
 * For accessing non-workload services like ssh, I would rather want to see websocket support like POCed in #1015

Where thinking about this may still be interesting is migrating VMs where they have additional networks which also have k8s endpoints configured. Not sure if the issue was for meant for that.

**Comment by fabiand** [comment]
My take would be to close this.

The similar behavior to the Kubernetes pod and service relationship is what makes sense for me here as well.
